src/include/migraphx/op/roialign.hpp

```diff
@@ -0,0 +1,321 @@
+#include <migraphx/shape_for_each.hpp>
+#include <cmath>
+#include <utility>
+#include <vector>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+namespace op {
+
+struct roialign
+{
+    std::string coord_trans_mode = "half_pixel";
+    std::string mode             = "avg";
+    int64_t output_height        = 1;
+    int64_t output_width         = 1;
+    int64_t sampling_ratio       = 0;
+    float spatial_scale          = 1.0f;
+
+    template <class Self, class F>
+    static auto reflect(Self& self, F f)
+    {
+        return pack(f(self.coord_trans_mode, "coordinate_transformation_mode"),
+                    f(self.mode, "mode"),
+                    f(self.output_height, "output_height"),
+                    f(self.output_width, "output_width"),
+                    f(self.sampling_ratio, "sampling_ratio"),
+                    f(self.spatial_scale, "spatial_scale"));
+    }
+
+    std::string name() const { return "roialign"; }
+
+    shape compute_shape(std::vector<shape> inputs) const
+    {
+        check_shapes{inputs, *this}.has(3).standard();
+        auto x_lens   = inputs.at(0).lens();
+        auto roi_lens = inputs.at(1).lens();
+        auto bi_lens  = inputs.at(2).lens();
+        auto type     = inputs.at(0).type();
+
+        // check input correct
+        if(bi_lens.size() != 1)
+        {
+            MIGRAPHX_THROW("ROIALIGN: batch indices should be 1 dimension!");
+        }
+
+        if(roi_lens.size() != 2 or roi_lens.at(1) != 4)
+        {
+            MIGRAPHX_THROW(
+                "ROIALIGN: rois should be 2 dimensions, and the second dim should be 4!");
+        }
+
+        if(roi_lens.front() != bi_lens.front())
+        {
+            MIGRAPHX_THROW("ROIALIGN: rois and batch indices inputs should have the same number!");
+        }
+
+        std::vector<std::size_t> out_lens = x_lens;
+        out_lens[0]                       = roi_lens[0];
+        out_lens[2]                       = output_height;
+        out_lens[3]                       = output_width;
+
+        return {type, out_lens};
+    }
+
+    struct pos_weight
+    {
+        int pos1;
+        int pos2;
+        int pos3;
+        int pos4;
+        float w1;
+        float w2;
+        float w3;
+        float w4;
+    };
+
+    void calc_pos_weight(const int64_t height,
+                         const int64_t width,
+                         const shape& comp_s,
+                         float roi_start_h,
+                         float roi_start_w,
+                         float bin_size_h,
+                         float bin_size_w,
+                         int64_t roi_bin_grid_h,
+                         int64_t roi_bin_grid_w,
+                         std::vector<pos_weight>& pos_weights) const
+    {
+        shape_for_each(comp_s, [&](auto idx) {
+            auto ph    = idx[0];
+            auto pw    = idx[1];
+            auto iy    = idx[2];
+            auto ix    = idx[3];
```

Comment:
These should be arrays as well.

Reply:
what does it mean? put "ph, pw, iy, ix" in an array?

Reply:
Well you would make two arrays:

```cpp
std::array<int64_t, 2> p = {idx[0}, idx[1]};
std::array<int64_t, 2> i = {idx[2], idx[3]};
```